src/common/lib/util/eventemitter.ts

```diff
@@ -286,32 +286,18 @@ class EventEmitter {
   }
 
   /**
-   * Private API
-   *
    * Listen for a single occurrence of a state event and fire immediately if currentState matches targetState
    * @param targetState the name of the state event to listen to
    * @param currentState the name of the current state of this object
-   * @param listener the listener to be called
-   * @param listenerArgs
    */
-  whenState(targetState: string, currentState: string, listener: Function, ...listenerArgs: unknown[]) {
-    const eventThis = { event: targetState };
-
+  async whenState(targetState: string, currentState: string) {
```

Comment:
After getting familiar with `whenState` API it doesn't feel right to me to have it as part of the EventEmitter class.
It just has too much business logic: why EventEmitter knows about the `state`; knows that state can change and knows what to do when target and current state are the same.

Not sure how we would change it though, only duplicating this code both in Connection and RealtimeChannel comes to mind and maybe at least adding the definition to the separate interface and making sure that Connection and RealtimeChannel implement that interface.

Probably can do it in a separate PR or just create a refactoring issue for now.

Reply:
Yeah, agree that `EventEmitter` isn't really the right place for it. Given that it's only an implementation detail (i.e. isn't in the public `EventEmitter` API), I don't think that it's very urgent to sort out. Created https://github.com/ably/ably-js/issues/1669 for it.